egui_plot/src/items/heatmap.rs

```diff
@@ -0,0 +1,469 @@
+    name: String,
+
+    /// Size of one tile in plot coordinates
+    tile_size: Vec2,
+}
+
+impl<const RESOLUTION: usize> PartialEq for Heatmap<RESOLUTION> {
+    /// manual implementation of `PartialEq` because formatter and color mapping do not impl `PartialEq`.
+    ///
+    /// > NOTE: custom_mapping and formatter are ignored
+    fn eq(&self, other: &Self) -> bool {
+        self.pos == other.pos
+            && self.values == other.values
+            && self.cols == other.cols
+            && self.rows == other.rows
+            && self.min == other.min
+            && self.max == other.max
+            && self.show_labels == other.show_labels
+            && self.palette == other.palette
+            && self.highlight == other.highlight
+            && self.name == other.name
+            && self.tile_size == other.tile_size
+    }
+}
+
+impl<const RESOLUTION: usize> Heatmap<RESOLUTION> {
+    /// Create a 2D heatmap. Will automatically infer number of rows.
+    ///
+    /// - `values` contains magnitude of each tile. The alignment is row by row.
+    /// - `cols` is the number of columns (i.e. the length of each row) and must not be zero.
+    /// - `values.len()` must be a multiple of `cols`.
+    ///
+    /// Example: To display this
+    ///
+    /// | -- | -- |
+    /// | 0.0 | 0.1 |
+    /// | 0.3 | 0.4 |
+    ///
+    /// pass `values = vec![0.0, 0.1, 0.3, 0.4]` and `cols = 2`.
+    ///
+    /// Returns error type if provided parameters are inconsistent
+    ///
+    /// # Errors
+    /// - `HeatmapErr::ZeroColumns` if `cols` is zero
+    /// - `HeatmapErr::ZeroRows` if `values` is empty
+    /// - `HeatmapErr::BadLength` if `values.len()` is not divisible by `cols`
+    pub fn new(values: Vec<f64>, cols: usize) -> Result<Self, HeatmapErr> {
+        if cols == 0 {
+            return Err(HeatmapErr::ZeroColumns);
+        }
+        if (values.len() % cols) != 0 {
+            return Err(HeatmapErr::BadLength);
+        }
+        let rows = values.len() / cols;
+        if rows == 0 {
+            return Err(HeatmapErr::ZeroRows);
+        }
+
+        // determine range
+        let mut min = f64::MAX;
+        let mut max = f64::MIN;
+        for v in &values {
+            min = min.min(*v);
+            max = max.max(*v);
+        }
+
+        Ok(Self {
+            base: PlotItemBase::new(String::new()),
+            pos: PlotPoint { x: 0.0, y: 0.0 },
+            values,
+            cols,
+            rows,
+            min,
+            max,
+            formatter: Box::new(|v| format!("{v:.1}")),
+            custom_mapping: None,
+            show_labels: true,
+            palette: Self::linear_gradient_from_base_colors(&BASE_COLORS),
+            highlight: false,
+            name: String::new(),
+            tile_size: Vec2 { x: 1.0, y: 1.0 },
+        })
+    }
+
+    /// Set color palette by specifying base colors from low to high
+    #[inline]
+    pub fn palette(mut self, base_colors: &[Color32]) -> Self {
```

Comment:
Just simply store the colors? If user wants finer-grained details, they can use the helper function.
Add it to docs comment as a helpful reference.